Add docs about rolling out TLS on k8s (#7096)
hashicorp/consul: website/source/docs/platform/k8s/upgrading.html.md

To initiate the upgrade, change the `server.image` value to the
desired Consul version. For illustrative purposes, the example below will

> use `consul:123.456`. Also, set the `server.updatePartition` value

Reviewer: now I'm always going to be unsure if this is your grammar skillz or grammarly's 😆

the clients, set the `client.image` value to the desired Consul version.
Then, run `helm upgrade`. This will upgrade the clients in batches, waiting
until the clients come up healthy before continuing.

> ## Configuring TLS on an Existing Cluster

Reviewer: I'd like to see this as a separate page. What do you think? You could link from here to the page.

Author: Yeah, I was on the fence about that, but could definitely split.

> As of version `0.16.0`, the Consul Helm chart supports TLS for communication

Reviewer: ```suggestion
As of Consul Helm version `0.16.0`, the chart supports TLS for communication between Consul components
```

within the cluster. If you already have a Consul cluster deployed on Kubernetes,
you may want to configure TLS in a way that minimizes downtime to your applications.
Consul already supports rolling out TLS on an existing cluster without downtime.
However, depending on your Kubernetes use case, your upgrade procedure may be different.

### Gradual TLS Rollout without Consul Connect

> If you're not using Consul Connect, follow this process.

Reviewer: ```suggestion
If you're **not using Consul Connect**, follow this process.
```

1. Run a Helm upgrade with the following config:
   ```yaml
   global:
     tls:
       enabled: true
       # This configuration sets `verify_outgoing`, `verfiy_server_hostname`,
       # and `verify_incoming` to `false` on servers and clients,
       # which allows TLS-disabled nodes to join the cluster.
       verify: false
   server:
     updatePartition: <number_of_server_replicas>
   ```
   This upgrade will trigger a rolling update of the clients, as well as any
   other `consul-k8s` components, such as sync catalog or client snapshot deployments.
1. Perform a rolling upgrade of the servers, as described in
   [Upgrade Consul Servers](#upgrading-consul-servers).
1. Repeat steps 1 and 2, turning on TLS verification by removing the

>    `global.tls.verify` property.

Reviewer: I'd say setting it to `true` since that makes more sense as a user

### Gradual TLS Rollout with Consul Connect

> Because the Envoy proxy needs to talk to the Consul client agent regularly

Reviewer: ```suggestion
Because the sidecar Envoy proxies need to talk to the Consul client agent regularly
```

> for service discovery, we can't enable TLS on the clients without also re-injecting

Reviewer: ```suggestion
for service discovery, we can't enable TLS on the clients without also re-injecting a
```

TLS-enabled proxy into the application pods. To perform TLS rollout with minimal
downtime, we recommend instead to add a new Kubernetes node pool and migrate your
applications to it.

1. Add a new identical node pool.

> 1. Cordon all nodes in the old pool by running `kubectl cordon`

Reviewer: ```suggestion
1. Cordon all nodes in the **old pool** by running `kubectl cordon`
```

>    to ensure Kubernetes doesn't schedule any new workloads on those nodes.

Reviewer: ```suggestion
   to ensure Kubernetes doesn't schedule any new workloads on those nodes and instead schedules onto the new nodes (which shortly will be TLS-enabled).
```

1. Create the following Helm config file for the upgrade:
    ```yaml
    global:
      tls:
        enabled: true
        # This configuration sets `verify_outgoing`, `verfiy_server_hostname`,
        # and `verify_incoming` to `false` on servers and clients,
        # which allows TLS-disabled nodes to join the cluster.
        verify: false
    server:
      updatePartition: <number_of_server_replicas>
    client:
      updateStrategy: |
         type: OnDelete
    ```
   In this configuration, we're setting `server.updatePartition` to the number of
   server replicas as described in [Upgrade Consul Servers](#upgrading-consul-servers)
   and `client.updateStrategy` to `OnDelete` to manually trigger an upgrade of the clients.

> 1. Run `helm upgrade` with the above config file. The upgrade will trigger an update of any

Reviewer: ```suggestion
1. Run `helm upgrade` with the above config file. The upgrade will trigger an update of all
```

>    component except clients and servers, such as Consul Connect webhook deployment

Reviewer: ```suggestion
   components except clients and servers, such as the Consul Connect webhook deployment
```

>    or sync catalog deployment. Note that the sync catalog and the client

Reviewer: ```suggestion
   or the sync catalog deployment. Note that the sync catalog and the client
```

   snapshot deployments will not be in the `ready` state until the clients on their
   nodes are upgraded. It is OK to proceed to the next step without them being ready
   because Kubernetes will keep the old deployment pod around, and so there will be no
   downtime.

> 1. Gradually perform an upgrade of the clients by deleting client pods on the new node

Reviewer: ```suggestion
1. Gradually perform an upgrade of the clients by deleting client pods on the **new** node
```

>    pool.

Reviewer: Can you add a new step:
```
1. At this point all components (e.g. Consul Connect webhook, sync catalog) should be running on the new node pool.
```

> 1. Redeploy all your Connect-enabled applications. Now that the Connect injector is TLS-aware,

Reviewer: We were calling it Connect Webhook above. Best stick to one name.

   it will add TLS configuration to the sidecar proxy. Also, Kubernetes should schedule

>    these applications on the new node pool.

Reviewer: Maybe add that one way to trigger a redeploy is to `kubectl drain` the nodes in the old pool.

1. Perform a rolling upgrade of the servers described in
   [Upgrade Consul Servers](#upgrading-consul-servers).
1. If everything is healthy, delete the old node pool.

> 1. Finally, remove `global.tls.verify` from your Helm config file, as well as

Reviewer: Set to true maybe?

>    `client.updateStrategy`, and perform a rolling upgrade of the servers.

Reviewer: Can you add a NOTE here at the bottom that it's possible to do this without fully duplicating the node pool via a rolling deploy: drain a subset of old servers, this is now your "new node pool", follow the rest of the steps and then repeat.

Don't need to go into too much detail but at least say it's possible.
